lib/membrane_rtc_engine/engine.ex

```diff
@@ -474,102 +473,66 @@ defmodule Membrane.RTC.Engine do
     {{:ok, tracks_msgs}, state}
   end
 
-  defp link_inbound_track({track_id, encoding}, tee, track_id_and_peer_to_status, ctx) do
-    links_tracks_peers =
+  defp link_inbound_track({track_id, encoding}, tee, awaiting_links, ctx) do
+    links_peers =
       flat_map_children(ctx, fn
         {:endpoint, other_endpoint_id} = other_endpoint_name ->
-          case Map.get(track_id_and_peer_to_status, {other_endpoint_id, track_id}) do
-            :waiting_for_linking ->
-              [
-                link(tee)
-                |> via_out(:copy)
-                |> via_in(Pad.ref(:input, track_id), options: [encoding: encoding])
-                |> to(other_endpoint_name)
-                |> then(fn linked -> {linked, track_id, other_endpoint_id} end)
-              ]
-
-            _other ->
-              []
+          tracks_set = Map.get(awaiting_links, other_endpoint_id)
+
+          if MapSet.member?(tracks_set, track_id) do
+            [
+              link(tee)
+              |> via_out(:copy)
+              |> via_in(Pad.ref(:input, track_id), options: [encoding: encoding])
+              |> to(other_endpoint_name)
+              |> then(fn linked -> {linked, other_endpoint_id} end)
+            ]
+          else
+            []
           end
 
         _child ->
           []
       end)
 
-    track_id_and_peer_to_status =
-      update_track_id_and_peer_to_status(
-        links_tracks_peers,
-        track_id_and_peer_to_status
-      )
+    awaiting_links =
+      Enum.map(links_peers, fn {_link, peer} -> peer end)
+      |> Enum.reduce(awaiting_links, fn
+        peer_id, acc ->
+          Map.update!(acc, peer_id, fn tracks_set -> MapSet.delete(tracks_set, track_id) end)
+      end)
 
-    Enum.map(links_tracks_peers, fn {link, _track_id, _peer_id} -> link end)
-    |> then(fn links -> {links, track_id_and_peer_to_status} end)
+    Enum.map(links_peers, fn {link, _peer_id} -> link end)
+    |> then(fn links -> {links, awaiting_links} end)
   end
```

Comment:
Try to write something like this
```elixir
  defp link_inbound_track({track_id, encoding}, tee, awaiting_links, ctx) do
    reduce_children(ctx, {[], awaiting_links}, fn
      {:endpoint, endpoint_id}, {new_links, awaiting_links} ->
        if MapSet.member?(awaiting_links[endpoint_id], track_id) do
          new_link =
            link(tee)
            |> via_out(:copy)
            |> via_in(Pad.ref(:input, track_id), options: [encoding: encoding])
            |> to({:endpoint, endpoint_id})

          awaiting_links = Map.update!(awaiting_links, endpoint_id, &MapSet.delete(&1, track_id))

          {new_links ++ [new_link], awaiting_links}
        else
          {new_links, awaiting_links}
        end

      _other_child, {new_links, awaiting_links} ->
        {new_links, awaiting_links}
    end)
  end
  ```
  
  To this end a new function similar to `flat_map_children` will be needed
  ```elixir
  defp reduce_children(ctx, acc, fun) do
    ctx.children |> Map.keys() |> Enum.reduce(acc, fun)
  end
  ```